interfaces/builtin/shared_memory.go

```diff
@@ -0,0 +1,155 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/snap"
+)
+
+const sharedMemorySummary = `allows two snaps to use predefined shared memory objects`
+
+const sharedMemoryBaseDeclarationSlots = `
+  shared-memory:
+    allow-installation:
+      slot-snap-type:
+        - app
+    allow-connection:
+      plug-attributes:
+        shared-memory: $SLOT(shared-memory)
+    allow-auto-connection:
+      plug-publisher-id:
+        - $SLOT_PUBLISHER_ID
+      plug-attributes:
+        shared-memory: $SLOT(shared-memory)
```

Comment:
I think we need to make this super-privileged for now since as I mentioned in the design doc, we don't have a way to know whether a snap is "actually" providing the file that is being granted access to, i.e. a snap may say it is providing access to /dev/shm/org.foo.secret by declaring that in the plug side, but really on the device /dev/shm/org.foo.secret is something that comes from outside snap confinement and the snaps can now access that even though the snaps are not creating the file. This is since /dev/shm is shared with the host filesystem.

Reply:
OK. Please check if the changes I made now are correct :-)